src/parser/pacbuild.rs

```diff
@@ -0,0 +1,2043 @@
+    pub ppa: Option<Vec<PPA>>,
+    pub repology: Option<Vec<RepologyFilter>>,
+    pub sources: Vec<Source>,
+
+    pub prepare: Option<String>,
+    pub build: Option<String>,
+    pub check: Option<String>,
+    pub package: Option<String>,
+    pub pre_install: Option<String>,
+    pub post_install: Option<String>,
+    pub pre_upgrade: Option<String>,
+    pub post_upgrade: Option<String>,
+    pub pre_remove: Option<String>,
+    pub post_remove: Option<String>,
+    pub custom_functions: Option<HashMap<String, String>>,
+}
+
+impl PacBuild {
+    fn cleanup_rawstring(raw_string: &str) -> &str {
+        let len = raw_string.len();
+        if len <= 2 {
+            ""
+        } else {
+            &raw_string[1..len - 1]
+        }
+    }
+
+    #[allow(clippy::too_many_lines)]
+    pub fn from_source(source_code: &str) -> Result<Self, ParseError> {
+        let mut pkgname: Option<Vec<Pkgname>> = None;
+        let mut pkgver: Option<PkgverType> = None;
+        let mut epoch: Option<u32> = None;
+        let mut pkgdesc: Option<String> = None;
+        let mut url: Option<String> = None;
+        let mut license: Option<Expression> = None;
+        let mut custom_variables: Option<HashMap<String, String>> = None;
+
+        let mut arch: Option<Vec<String>> = None;
+        let mut maintainer: Option<Vec<Maintainer>> = None;
+        let mut noextract: Option<Vec<String>> = None;
+        let mut sha256sums: Option<HashMap<String, Vec<Option<String>>>> = None;
+        let mut sha348sums: Option<HashMap<String, Vec<Option<String>>>> = None;
+        let mut sha512sums: Option<HashMap<String, Vec<Option<String>>>> = None;
+        let mut b2sums: Option<HashMap<String, Vec<Option<String>>>> = None;
+        let mut depends: Option<Vec<Dependency>> = None;
+        let mut optdepends: Option<Vec<OptionalDependency>> = None;
+        let mut ppa: Option<Vec<PPA>> = None;
+        let mut repology: Option<Vec<RepologyFilter>> = None;
+        let mut sources: Option<Vec<Source>> = None;
+
+        let mut prepare: Option<String> = None;
+        let mut build: Option<String> = None;
+        let mut check: Option<String> = None;
+        let mut package: Option<String> = None;
+        let mut pre_install: Option<String> = None;
+        let mut post_install: Option<String> = None;
+        let mut pre_upgrade: Option<String> = None;
+        let mut post_upgrade: Option<String> = None;
+        let mut pre_remove: Option<String> = None;
+        let mut post_remove: Option<String> = None;
+        let mut custom_functions: Option<HashMap<String, String>> = None;
+
+        #[allow(clippy::similar_names)]
+        let sourced_code = Command::new("bash")
+            .args([
+                "-c",
+                &format!(r#"{source_code}; SOURCED_CODE="$(declare -p | cut -d ' ' -f 3-)"; TAIL="$(echo "$SOURCED_CODE" | grep -m 1 -n '_=\"\"' | cut -d ':' -f 1)"; echo "$SOURCED_CODE" | tail -n +$(($TAIL + 1)); declare -f"#),
+            ])
```

Comment:
Isolate shell context and filter bash builtins instead of blanket filtering uppercase
```suggestion
        let sourced_code = Command::new("env")
            .args([
                "-i",
                "bash",
                "-c",
                &format!(r#"VARS="$(compgen -v)"; {source_code}; VARS="`grep -vFe "$VARS" <<< "$(compgen -v)"|grep -v VARS |grep -v PIPESTATUS`"; for v in $VARS;do declare -p $v|cut -d " " -f 3-; done; declare -f"#),
            ])
```